python/ray/train/lightning/lightning_trainer.py

```diff
@@ -0,0 +1,374 @@
+            transformed by the ``preprocessor`` if one is provided.
+        datasets_iter_config: Configurations for iterating over input Ray datasets.
+            This configuration is only valid when `datasets` argument is provided to
+            the LightningTrainer. Otherwise, LightningTrainer will use datamodule
+            or dataloaders specified in ``LightningConfig.trainer_init_config``.
+            For valid arguments to pass, please refer to:
+            :py:meth:`Dataset.iter_torch_batches <ray.data.Dataset.iter_torch_batches>`
+        preprocessor: A ray.data.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        lightning_config: Optional[Dict[str, Any]] = None,
+        *,
+        torch_config: Optional[TorchConfig] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        run_config: Optional[RunConfig] = None,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        datasets_iter_config: Optional[Dict[str, Any]] = None,
+        preprocessor: Optional[Preprocessor] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+        train_loop_config = {
+            "lightning_config": lightning_config,
+            "datasets_iter_config": datasets_iter_config,
+        }
+
+        super(LightningTrainer, self).__init__(
+            train_loop_per_worker=_lightning_train_loop_per_worker,
+            train_loop_config=train_loop_config,
+            torch_config=torch_config,
+            scaling_config=scaling_config,
+            dataset_config=dataset_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+
+def _lightning_train_loop_per_worker(config):
+    """Per-worker training loop for a Lightning Trainer."""
+    if not config["lightning_config"]:
+        raise RuntimeError("'lightning_config' not specified in LightningTrainer!")
+
+    # Unpack all configs
+    ptl_config = config["lightning_config"]
+    datasets_iter_config = config["datasets_iter_config"]
+    trainer_config = ptl_config["_trainer_init_config"]
+    trainer_fit_params = ptl_config["_trainer_fit_params"]
+    module_class = ptl_config["_module_class"]
+    module_init_config = ptl_config["_module_init_config"]
+    ddp_strategy_config = ptl_config["_ddp_strategy_config"]
+
+    # Prepare data
+    datamodule = trainer_fit_params.get("datamodule", None)
+    train_dataloaders = trainer_fit_params.get("train_dataloaders", None)
+    val_dataloaders = trainer_fit_params.get("val_dataloaders", None)
+
+    train_ray_dataset = session.get_dataset_shard("train")
+    val_ray_dataset = session.get_dataset_shard("val")
```

Comment:
What happens if they provide both `train_dataloaders` and ray datasets? We should log a warning that the ray datasets will be used.

Reply:
Same thing for if they supply both `train_dataloaders` and `datamodule`. or all 3

Reply:
Lightning will throw an error if users provide both `dataloaders` and `datamodules`. If they provide both dataloaders and Ray datasets, trainer.fit() will fail.

The only circumstance we need to care about is user provide both datamodules and Ray datasets. I already raise a warning in line 300.

Reply:
Added a check to ensure at least one data source is provided.